clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -636,42 +857,148 @@ public void assign(Collection<TopicPartition> partitions) {
+
+        event.future().whenComplete((completedFetches, error) -> {
+            if (completedFetches != null && !completedFetches.isEmpty()) {
+                fetchBuffer.addAll(completedFetches);
+            }
+        });
+    }
+
+    /**
+     * @throws KafkaException if the rebalance callback throws exception
+     */
+    private Fetch<K, V> pollForFetches(Timer timer) {
+        long pollTimeout = timer.remainingMs();
+
+        // if data is available already, return it immediately
+        final Fetch<K, V> fetch = fetchCollector.collectFetch(fetchBuffer);
+        if (!fetch.isEmpty()) {
+            return fetch;
+        }
+
+        // send any new fetches (won't resend pending fetches)
+        sendFetches();
+
+        // We do not want to be stuck blocking in poll if we are missing some positions
+        // since the offset lookup may be backing off after a failure
+
+        // NOTE: the use of cachedSubscriptionHasAllFetchPositions means we MUST call
+        // updateAssignmentMetadataIfNeeded before this method.
+        if (!cachedSubscriptionHasAllFetchPositions && pollTimeout > retryBackoffMs) {
+            pollTimeout = retryBackoffMs;
+        }
+
+        log.trace("Polling for fetches with timeout {}", pollTimeout);
+
+        Timer pollTimer = time.timer(pollTimeout);
+
+        // Attempt to fetch any data. It's OK if we time out here; it's a best case effort. The
+        // data may not be immediately available, but the calling method (poll) will correctly
+        // handle the overall timeout.
+        try {
+            Queue<CompletedFetch> completedFetches = eventHandler.addAndGet(new FetchEvent(), pollTimer);
+            if (completedFetches != null && !completedFetches.isEmpty()) {
+                fetchBuffer.addAll(completedFetches);
+            }
+        } catch (TimeoutException e) {
+            log.trace("Timeout during fetch", e);
+        } finally {
+            timer.update(pollTimer.currentTimeMs());
+        }
 
-        clusterResourceListeners.maybeAdd(keyDeserializer);
-        clusterResourceListeners.maybeAdd(valueDeserializer);
-        return clusterResourceListeners;
+        return fetchCollector.collectFetch(fetchBuffer);
+    }
+
+    /**
+     * Set the fetch position to the committed position (if there is one)
+     * or reset it using the offset reset policy the user has configured.
+     *
+     * @throws org.apache.kafka.common.errors.AuthenticationException if authentication fails. See the exception for more details
+     * @throws NoOffsetForPartitionException If no offset is stored for a given partition and no offset reset policy is
+     *             defined
+     * @return true iff the operation completed without timing out
+     */
+    private boolean updateFetchPositions(final Timer timer) {
+        // If any partitions have been truncated due to a leader change, we need to validate the offsets
+        eventHandler.add(new ValidatePositionsApplicationEvent());
```

Comment:
My understanding is that in that case it won't continuously fetch from the old leader because of the partition subscription state, which will transition to `AWAITING_VALIDATION` (not a valid state for fetching). The moment a new metadata is discovered, [partitions are marked as requiring validation](https://github.com/apache/kafka/blob/59d0daf247a066ad408f713dbfb19e92fc243200/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java#L195), so the [prepareFetchRequest](https://github.com/apache/kafka/blob/59d0daf247a066ad408f713dbfb19e92fc243200/clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractFetch.java#L376C12-L376C12) will not fetch from them. 

To complete the validation picture, once the `OffsetForLeaderEpoch` response is received, the subscription state will transition to `FETCHING` again and fetch requests should resume for that partition. 

My thoughts but please @kirktrue correct me if I'm missing something from the fetching story.

Reply:
Yes, it appears that it will continue to attempt to fetch from the old leader.

When a fetch response is received from the broker and we notice that it has an error, we call `FetchCollector.handleInitializeErrors()` to deal with the different error conditions. When it notices that the error is `FENCED_LEADER_EPOCH`, it will execute these two statements (from the `FetchUtils.requestMetadataUpdate()` method):

```java
metadata.requestUpdate(false);
subscriptions.clearPreferredReadReplica(topicPartition);
```

That's all it does. The logic doesn't update the `FetchState` for that partition. It doesn't clear out the leader epoch. Nothing.

The next time the user calls the `Consumer.poll()` method, the `Fetcher`/`FetchRequestManager` will determine for which partitions we should issue `FETCH` RPCs. The first step is to call the `SubscriptionState.fetchablePartitions()` method which checks each partition:

```java
public synchronized List<TopicPartition> fetchablePartitions(Predicate<TopicPartition> isAvailable) {
    // Since this is in the hot-path for fetching, we do this instead of using java.util.stream API
    List<TopicPartition> result = new ArrayList<>();
    assignment.forEach((topicPartition, topicPartitionState) -> {
        // Cheap check is first to avoid evaluating the predicate if possible
        if (topicPartitionState.isFetchable() && isAvailable.test(topicPartition)) {
            result.add(topicPartition);
        }
    });
    return result;
}
```

Because we didn't change anything in the underlying state of the partition in `SubscriptionState` previously, when the `topicPartitionState.isFetchable()` method is invoked, it returns `true`. Thus it is included in the list for which to fetch, and we will likely hit the same error.

Reply:
The error message in `Errors` for `FENCED_LEADER_EXCEPTION` states that this error is caused when...

> The leader epoch in the request is older than the epoch on the broker.

Reply:
Naively it seems like we want to clear/reset the partition's `Metadata.LeaderAndEpoch` value in the metadata cache locally. Then in the fetch logic, that would allow us to skip that partition when we check for the presence of a leader:

```java
for (TopicPartition partition : fetchablePartitions()) {
    SubscriptionState.FetchPosition position = subscriptions.position(partition);

    if (position == null)
        throw new IllegalStateException("Missing position for fetchable partition " + partition);

    Optional<Node> leaderOpt = position.currentLeader.leader;

    if (!leaderOpt.isPresent()) {
        log.debug("Requesting metadata update for partition {} since the position {} " +
                  "is missing the current leader node", partition, position);
        metadata.requestUpdate(false);
        continue;
    }

    . . .
}
```

Reply:
Thanks, Kirk. Does the old consumer have the same behavior too? Should we file a followup jira to improve this?

Reply:
I will triple check, and if so, I will add a Jira.

Reply:
I have filed KAFKA-15615 to resolve.